Report NamespaceNotFederated in propagation status

marun opened the pull request:
Previously a namespaced resource whose containing namespace was not federated (i.e. did not contain a `FederatedNamespace` resource) would report a healthy propagation status.  This has been changed to reporting unhealthy propagation status with reason `NamespaceNotFederated`.

Fixes #951

marun commented:
CI failure was a flake of the walkthrough test (#952), and tests affected by this PR appear to be otherwise passing.

marun commented:
I need to update the test to wait for the condition.

marun commented:
CI failure due to #962.  I don't think this PR is the cause.

xunpan commented:
LGTM

marun commented:
Rebased

marun commented:
Rebased on #973

marun commented:
Rebased and ready for review again

font commented on test/common/crudtester.go:
```diff
@@ -739,3 +731,25 @@ func (c *FederatedTypeCrudTester) CheckStatusCreated(qualifiedName util.Qualifie
+
+	// Convert the resource to the status interface
+	genericStatus := &status.GenericFederatedStatus{}
+	err = util.UnstructuredToInterface(fedObject, genericStatus)
```
How does `fedObject`, which is the unstructured version of a `FederatedX` resource, get unmarshalled to a `GenericFederatedStatus` struct? Don't we need to be working with the `status` field from the unstructured version of `FederatedX` and the `GenericFederatedStatus` struct?

marun replied:
Take a look at `GenericFederatedStatus` - it has an optional `Status` field.  Unmarshalling from the unstructured resource to the struct will populate the `Status` field in the struct if it is present in the resource.

marun commented on pkg/controller/sync/controller.go:
```diff
@@ -90,6 +90,8 @@ type KubeFedSyncController struct {
 
 	skipAdoptingResources bool
+
+	limitedScope bool
```
I much prefer a boolean for internal use, where extensibility can be achieved without having to complicate the code with  enumerations.

marun commented on pkg/controller/sync/status/status.go:
```diff
@@ -65,6 +65,7 @@ const (
 	ClusterRetrievalFailed AggregateReason = "ClusterRetrievalFailed"
 	ComputePlacementFailed AggregateReason = "ComputePlacementFailed"
 	CheckClusters          AggregateReason = "CheckClusters"
+	NamespaceNotFederated  AggregateReason = "NamespaceNotFederated"
```
Good catch, `AggregateStatus` is no removed.  I would have thought the linter would catch that sort of thing.

Also removed the pointer usage for `GenericCondition`, I'm not sure where that came from.

marun commented on pkg/controller/sync/controller.go:
```diff
@@ -390,6 +393,13 @@ func (s *KubeFedSyncController) setPropagationStatus(fedResource FederatedResour
+	// For a cluster-scoped control plane, report when the containing namespace of a federated
+	// resource is not federated.  The KubeFed system namespace is implicitly federated in a
+	// namespace-scoped control plane.
+	if !s.limitedScope && fedResource.NamespaceNotFederated() && reason == status.AggregateSuccess {
```
I think allowing more than one error state per cluster would be more confusing than helpful, so `NamespaceNotFederated` is indicated only if no other error state has been recorded (i.e. `reason == status.AggregateSuccess`).  I've reworked this block and added a comment to hopefully make this clear.

Regarding your second question, only the `CheckClusters` reason is dependent on `PropagationStatus` and that is determined in the call to `SetPropagationStatus` below.

marun commented on test/common/crudtester.go:
```diff
@@ -739,3 +731,25 @@ func (c *FederatedTypeCrudTester) CheckStatusCreated(qualifiedName util.Qualifie
+		return nil, errors.Wrapf(err, "Failed to retrieve federated resource from the API")
+	}
+
+	// Convert the resource to the status interface
```
Updated to specify struct instead of interface.  The previous use of the term 'interface' was related to  `UnstructuredToInterface`, which unmarshalls to a variable of type `interface{}`.

marun commented:
Updated to address comments.  I'm using fixup commits to address comments with the intention of squashing before merge, and have accordingly put a hold on the PR to avoid merging before squash.

font commented:
Thanks @marun! LGTM

font commented:
Also looks like a legitimate CI failure.

marun commented:
Rebased and addressed the test failure. 

Also, I remember why I like using struct pointers with slices. This message has brought to you by the winning combination of pass-by-value and `append`.

marun commented:
Rebased

marun commented:
@font I've stripped this down to the bugfix.  I'll submit unrelated cleanup separately to avoid blocking this further at this time.

marun commented:
Rebased

font commented:
@marun This needs a conflict resolution.

shashidharatd commented:
LGTM. @marun please rebase. Thanks!

marun commented:
Rebased

font commented:
/lgtm
